File: examples/src/image_preprocessing_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import String
import time


class Image_preprocessingNode(Node):
    """Image_preprocessing stage node for vision_pipeline pipeline."""
    
    def __init__(self):
        super().__init__('image_preprocessing_node')
        
        # Set namespace
        self.set_namespace('/vision_pipeline/image_preprocessing')
        
        self.get_logger().info(f'Initializing Image_preprocessingNode')
        
        # Initialize state
        self.initialized = False
        self.processing_count = 0
        
        # Declare parameters
        self.declare_parameter('preprocessing_status_topic', '/pipeline/preprocessing_status')
        self.preprocessing_status_topic = self.get_parameter('preprocessing_status_topic').value
        
        # Initialize publishers
        self./pipeline/preprocessed_publisher = self.create_publisher(
            String, '/vision_pipeline/image_preprocessing//pipeline/preprocessed', 10)
        
        # Initialize subscribers
        self./camera/image_raw_subscriber = self.create_subscription(
            String, '/vision_pipeline/image_preprocessing//camera/image_raw',
            self.on_/camera/image_raw_received, 10)
        
        # Initialize timer for periodic processing
        self.timer = self.create_timer(0.1, self.timer_callback)  # 10 Hz
        
        self.initialized = True
        self.get_logger().info('Image_preprocessingNode initialized successfully')
    
    def on_/camera/image_raw_received(self, msg):
        """Handle received /camera/image_raw message."""
        self.get_logger().info(f'Received /camera/image_raw: {msg.data}')
        
        # Process the input
        self.preprocess_image()
        
        # Publish outputs if available
        output_msg = String()
        output_msg.data = f'Processed /camera/image_raw: {msg.data}'
        self./pipeline/preprocessed_publisher.publish(output_msg)
    
    def preprocess_image(self):
        """Execute preprocess_image processing."""
        self.get_logger().debug(f'Executing preprocess_image')
        
        # TODO: Implement preprocess_image logic
        # This is where the actual processing logic would go
        
        self.processing_count += 1
    
    def timer_callback(self):
        """Periodic timer callback."""
        if not self.initialized:
            return
        
        # Periodic processing
        self.get_logger().debug(f'Image_preprocessingNode timer callback, count: {self.processing_count}')
        
        # Publish periodic status
        status_msg = String()
        status_msg.data = f'Image_preprocessingNode status: {self.processing_count}'
        self./pipeline/preprocessed_publisher.publish(status_msg)


def main(args=None):
    rclpy.init(args=args)
    
    node = Image_preprocessingNode()
    
    node.get_logger().info('Starting Image_preprocessingNode')
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main() 

File: examples/src/image_preprocessing_node.cpp
#include "robodsl_project/image_preprocessing_node.hpp"
#include <chrono>
#include <iostream>

using namespace std::chrono_literals;

namespace robodsl_project {

image_preprocessingNode::image_preprocessingNode() 
    : Node("image_preprocessing_node", "/vision_pipeline/image_preprocessing") {
    
    RCLCPP_INFO(this->get_logger(), "Initializing image_preprocessing stage node");
    
    // Initialize ROS components
    initialize_ros_components();
    
    
    
    RCLCPP_INFO(this->get_logger(), "image_preprocessing stage node initialized successfully");
}

image_preprocessingNode::~image_preprocessingNode() {
    
}

void image_preprocessingNode::initialize_ros_components() {
    // Initialize subscribers
    /camera/image_raw_sub_ = this->create_subscription<std_msgs::msg::String>(
        "/vision_pipeline/image_preprocessing//camera/image_raw", 10,
        std::bind(&image_preprocessingNode::/camera/image_raw_callback, this, std::placeholders::_1)
    );
    
    // Initialize publishers
    /pipeline/preprocessed_pub_ = this->create_publisher<std_msgs::msg::String>(
        "/vision_pipeline/image_preprocessing//pipeline/preprocessed", 10
    );
    
    // Initialize timer for periodic processing
    timer_ = this->create_wall_timer(
        std::chrono::milliseconds(100),
        std::bind(&image_preprocessingNode::timer_callback, this)
    );
}



// Callback methods
void image_preprocessingNode::/camera/image_raw_callback(const std_msgs::msg::String::SharedPtr msg) {
    RCLCPP_INFO(this->get_logger(), "Received /camera/image_raw: %s", msg->data.c_str());
    
    // Process the received data
    // Simple processing without CUDA/ONNX
    auto output_msg = std_msgs::msg::String();
    output_msg.data = "Processed: " + msg->data;
    /pipeline/preprocessed_pub_->publish(output_msg);
}

void image_preprocessingNode::timer_callback() {
    // Periodic processing tasks
    preprocess_image();
    
}

// Processing methods
void image_preprocessingNode::preprocess_image() {
    RCLCPP_DEBUG(this->get_logger(), "Executing preprocess_image");
    // TODO: Implement preprocess_image logic
}

// Model processing methods

} // namespace robodsl_project

#include "rclcpp_components/register_node_macro.hpp"
RCLCPP_COMPONENTS_REGISTER_NODE(robodsl_project::image_preprocessingNode) 